component/attr_cache/cacheMap.go

```diff
@@ -55,39 +58,102 @@ func newAttrCacheItem(attr *internal.ObjAttr, exists bool, cachedAt time.Time) *
+	if attr == nil {
+		return nil
+	}
+	path := internal.TruncateDirName(attr.Path)
+	//start recursion
+	cachedItem := value.insertHelper(attr, exists, cachedAt, path, "")
+	return cachedItem
+}
+
+// TODO: write unit tests for this
+func (value *attrCacheItem) insertHelper(attr *internal.ObjAttr, exists bool, cachedAt time.Time, path string, itemPath string) *attrCacheItem {
+	var cachedItem *attrCacheItem
+	paths := strings.SplitN(path, "/", 2) // paths[0] is home paths[1] is user/folder/file
+	if value.children == nil {
+		value.children = make(map[string]*attrCacheItem)
+	}
+	if len(paths) < 2 {
+		// this is a leaf
+		cachedItem = newAttrCacheItem(attr, exists, cachedAt)
+		value.children[paths[0]] = cachedItem
+	} else {
+		itemPath += paths[0] + "/"
+		//see if the directory exists. if not, create it.
+		_, ok := value.children[paths[0]]
+		if !ok {
+			value.children[paths[0]] = newAttrCacheItem(internal.CreateObjAttrDir(itemPath), exists, cachedAt)
+		}
+		cachedItem = value.children[paths[0]].insertHelper(attr, exists, cachedAt, paths[1], itemPath)
+	}
+	return cachedItem
+}
+
+// input: full path to item or file as string
+// output: the attrCacheItem value for the key found in path
+// description: a lookup of any attrCacheItem based on any given full path.
+func (value *attrCacheItem) get(path string) (*attrCacheItem, error) {
+	path = internal.TruncateDirName(path)
+	paths := strings.Split(path, "/")
+	currentItem := value
+	for _, pathElement := range paths {
+		//check if we are at the last element in the paths list
+		if path == "" {
+			break
+		}
+		var ok bool
+		currentItem, ok = currentItem.children[pathElement]
+		//check to see if directory (pathElement) exists
+		if !ok {
+			return nil, fmt.Errorf("The path element : %s does not exist", pathElement)
+		}
+		//TODO: side note: cacheLocks. channel, sync, semiphore.
+	}
+	return currentItem, nil
+}
+
 func (value *attrCacheItem) valid() bool {
 	return value.attrFlag.IsSet(AttrFlagValid)
 }
 
 func (value *attrCacheItem) exists() bool {
-	return value.attrFlag.IsSet(AttrFlagExists)
+	return value.valid() && value.attrFlag.IsSet(AttrFlagExists)
 }
 
+// TODO: don't return true for deleted files.
 func (value *attrCacheItem) isInCloud() bool {
 	isObject := !value.attr.IsDir()
 	isDirInCloud := value.attr.IsDir() && !value.attrFlag.IsSet(AttrFlagNotInCloud)
 	return isObject || isDirInCloud
 }
 
 func (value *attrCacheItem) markDeleted(deletedTime time.Time) {
-	value.attrFlag.Clear(AttrFlagExists)
-	value.attrFlag.Set(AttrFlagValid)
-	value.cachedAt = deletedTime
-	value.attr = &internal.ObjAttr{}
+	if !value.isDeleted() {
+		value.attrFlag.Clear(AttrFlagExists)
+		value.attrFlag.Set(AttrFlagValid)
+		value.cachedAt = deletedTime
+		value.attr = &internal.ObjAttr{}
+		for _, val := range value.children {
+			val.markDeleted(deletedTime)
+		}
+	}
 }
 
 func (value *attrCacheItem) invalidate() {
 	value.attrFlag.Clear(AttrFlagValid)
```

Comment:
Similar to markDeleted, you could save a bit of time by checking if the entry is already invalid before recursing over the children.